src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -1305,55 +1252,77 @@ private static int GetWarningsAsErrorOrMessageKey(BuildEventArgs buildEventArgs)
         /// </summary>
-        private void CreateLoggingEventQueue()
+        private void StartLoggingEventProcessing()
         {
-            // We are creating a two-node dataflow graph here.  The first node is a buffer, which will hold up to the number of
-            // logging events we have specified as the queueCapacity.  The second node is the processor which will actually process each message.
-            // When the capacity of the buffer is reached, further attempts to send messages to it will block.
-            // The reason we can't just set the BoundedCapacity on the processing block is that ActionBlock has some weird behavior
-            // when the queue capacity is reached.  Specifically, it will block new messages from being processed until it has
-            // entirely drained its input queue, as opposed to letting new ones in as old ones are processed.  This is logged as 
-            // a perf bug (305575) against Dataflow.  If they choose to fix it, we can eliminate the buffer node from the graph.
-            var dataBlockOptions = new DataflowBlockOptions
-            {
-                BoundedCapacity = Convert.ToInt32(_queueCapacity)
-            };
-
-            var loggingQueue = new BufferBlock<object>(dataBlockOptions);
-
-            var executionDataBlockOptions = new ExecutionDataflowBlockOptions
+            _eventQueue = new ConcurrentQueue<object>();
+            _dequeueEvent = new AutoResetEvent(false);
+            _emptyQueueEvent = new AutoResetEvent(false);
+            _enqueueEvent = new AutoResetEvent(false);
+            _loggingEventProcessingCancellation = new CancellationTokenSource();
+
+            _loggingEventProcessingThread = new Thread(LoggingEventProc);
+            _loggingEventProcessingThread.Name = $"MSBuild LoggingService events queue pump: {this.GetHashCode()}";
+            _loggingEventProcessingThread.IsBackground = true;
+            _loggingEventProcessingThread.Start();
+
+            void LoggingEventProc()
             {
-                BoundedCapacity = 1
-            };
-
-            var loggingQueueProcessor = new ActionBlock<object>(loggingEvent => LoggingEventProcessor(loggingEvent), executionDataBlockOptions);
+                var completeAdding = _loggingEventProcessingCancellation.Token;
 
-            var dataLinkOptions = new DataflowLinkOptions
-            {
-                PropagateCompletion = true
-            };
+                do
+                {
+                    // We peak message first in order to not have _eventQueue.IsEmpty before we actually process event
+                    //   as this could be interpreted like "every message has been already processed" otherwise.
+                    if (_eventQueue.TryPeek(out object ev))
+                    {
+                        LoggingEventProcessor(ev);
+                        _eventQueue.TryDequeue(out _);
+                        _dequeueEvent.Set();
+                    }
```

Comment:
If you have multiple competing TryPeeks, that would mean you could process a single event twice, right? Should this be:
```suggestion
                   if (_eventQueue.TryDequeue(out object ev) {
                        LoggingEventProcessor(ev);
                        _dequeueEvent.Set();
                    }
```

?

Reply:
That is true. But we don't have multiple competing `TryPeek` and chance we will need it in future is slim to none.
As stated in comment with TryPeek, process, Dequeue wee achieve that queue is empty only after all messages has been processed by listeners/loggers. 
Method `WaitForLoggingToProcessEvents` is used in places where logic requires all messages to be fully handled before continue and its implementation cosider job to be done when `queue.IsEmpty`.

Reply:
Ok; I think that makes sense.